### PR TITLE
Use a simpler way to delete recursive function

### PR DESCRIPTION
**Pull request by Stonesjtu**

### CODE REVIEW

File: torch/nn/parallel/scatter_gather.py [hunk 1]
```diff
@@ -24,10 +24,9 @@ def scatter_map(obj):
     # to a closure that has a reference to the scatter_map cell (because the
     # fn is recursive). To avoid this reference cycle, we set the function to
     # None, clearing the cell
-    try:
-        return scatter_map(inputs)
-    finally:
-        scatter_map = None
+    res = scatter_map(inputs)
```
Comment (mrshenli) [inline torch/nn/parallel/scatter_gather.py]: What if `scatter_map` throws an exception, will this still be able to clear the reference cycle?

**Comment by Stonesjtu** [comment]
Absolutely not. But I don't think we should manage such a corner case
because I don't think people are going to catch the error of gather/scatter
and continues to training. So once exception is raised, the whole process
is going to be terminated.

But if such case happens a lot, I think wrapping the function call into a
try/catch block is ok.

The major concern for this PR is readability.

**Comment by mrshenli** [comment]
@Stonesjtu Scatter and gather are used by `DataParallel` to distribute inputs and collect outputs. I saw cases where people catch CUDA OOM error from `DataParallel` in case the input they passed to `DataParallel` is too large, and then adjust and try again, instead of blowing up the entire training loop. If the PR cannot clear the reference cycle, I am inclined to keep the try-finally block. @pietern @ezyang thoughts?

**Comment by Stonesjtu** [comment]
Well, then I would prefer placing the `res = scatter_map(inputs` into a try
block, instead of the current one.

**Comment by mrshenli** [comment]
Sure, if you can make the changes accordingly, I will approve and land. @Stonesjtu

**Review by mrshenli — APPROVED** [review]
Thanks for contributing!

**Pull request closed without merging**